### PR TITLE
Vibrational averaging inputs and Fchk output parsers

### PR DESCRIPTION
**Pull request by herbertludowieg**
This is the first step to address the enhancement #128.
Additions:

- va.va.py
1. Generators for vibrational averaging input files and calculation of displacements in positive and negative directions with a specified delta for displacements as defined in the paper on #128. Limited to Gaussian script generation. I am aware that there already exists a Gaussian input generator and I would like to utilize the existing generator at a later date. This was done since I needed a very specific file writing procedure and have not been able to see if I can generalize it enough to utilize the existing functions.
2. to_va and write_grad_prop functions are temporary and only exist to be able to compare to existing program for Vibrational Averaging.

- gaussian.output.Fchk
1. Added parse_frequency, parse_frequency_ext, parse_shielding_tensor and parse_gradient functions to get the respective data from the FChk files.
2. Added a tolerance value so we can set small values read from the file to be 0.

- core.gradient
1. New file to define gradient dataframe. Still unsure if this is necessary.

Looking ahead:
Need to add tests for the new functions.
Generalize input generator so as to encompass more than one quantum chemistry package.

### CODE REVIEW

File: docs/source/api/interfaces.rst [hunk 1]
```diff
@@ -13,8 +13,5 @@ External Interfaces
 .. automodule:: exatomic.interfaces.pdb
     :members:
 
-.. automodule:: exatomic.interfaces.tens
-    :members:
-
 .. automodule:: exatomic.interfaces.xyz
```
Comment (avmarchenko) [inline docs/source/api/interfaces.rst]: Thanks for considering documentation! We should migrate to generating documentation using sphinx-apidoc (programmatic generation of rst files). Since that is outside the scope of this PR I've made an issue for this.

File: exatomic/core/gradient.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import numpy as np
+import pandas as pd
+
+class Gradient(DataFrame):
+    """
+    The gradient dataframe
+    """
+    # simple function that will have to be seen if it can have any other functions
+    _index = 'gradient'
+    _columns = ['Z', 'label', 'fx', 'fy', 'fz', 'symbol', 'frame']
```
Comment (avmarchenko) [inline exatomic/core/gradient.py]: To have this dataframe link to a specific atom, columns must require the atom index (`atom`)

File: exatomic/gaussian/output.py [hunk 3]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+        _redisp = 'Vib-Modes'
+        _refinfo = 'Vib-E2'
+        _remass = 'Vib-AtMass'
+        # TODO: find out what these two values are useful for
+        _rendim = 'Vib-NDim'
+        _rendim0 = 'Vib-NDim0'
+
+        found = self.find(_renmode)
+        if not found:
+            return
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: There might be a benefit to setting an empty `self.frequency_ext` here prior to exiting. Depends on what the desired behavior is in the case that this parse method fails (doesn't find anything to parse). If the attribute isn't set here, any logic that expects an instance of this class to have the `frequency_ext` attribute will throw an exception. I'm open to ideas/reasons on how to handle this (including maintaining it this way).

File: exatomic/gaussian/output.py [hunk 4]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+        else:
+            found = self.find(_renmode, _redisp, _refinfo, _remass, keys_only=True)
+        # find number of vibrational modes
+        nmode = self._intme(found[_renmode])
+        # get extended data (given by mapper array)
+        all_info = self._dfme(found[_refinfo], nmode * 14)
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: I think this parsing logic is a bit fragile. If any of the strings (e.g. `_renmode`) are not found a `KeyError` will be thrown. I think it might be worth refactoring this parsing logic to early exit if a key is missing. See also the above comment.

File: exatomic/gaussian/output.py [hunk 5]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+        all_info[abs(all_info) < self.tol] = 0
+        # mapper array to filter through all of the values printed on the fchk file
+        # TODO: have to find labels of unknown columns
+        mapper = ["freq", "r_mass", "f_const", "ir_int", "unk1", "unk2", "unk3",
+                  "dipole_s", "rot_s", "unk4", "unk5", "unk6", "unk7", "em_angle"]
+        # build extended frequency table
+        self.frequency_ext = pd.DataFrame.from_dict({mapper[int(i/nmode)]: all_info[i:i+nmode]
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: Would integer division work here (`//`) instead of the cast to int? Arguably cleaner if it does.

File: exatomic/gaussian/output.py [hunk 6]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+            return
+        else:
+            found = self.find(_renmode, _redisp, keys_only=True)
+        # get atomic numbers
+        znums = self.atom['Zeff'].values
+        # get number of atoms
+        nat = len(znums)
+        # TODO: will the order of the atoms and their frequencies always be the same?
+        # get frequency modes from extended data table
+        freq = self.frequency_ext['freq']
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: Since the above parse function has the possibility of returning without setting this attribute, this has the potential to throw an `AttributeError`. I think this should be handled somehow.

File: exatomic/gaussian/output.py [hunk 7]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+        x = shield[::3]
+        y = shield[1::3]
+        z = shield[2::3]
+#        # conversion from Hz to ppm
+#        # (only done for some test calculations may not be accurate for all)
+#        x *= 1e6/18778.86
+#        y *= 1e6/18778.86
+#        z *= 1e6/18778.86
+        matrix = np.transpose([x,y,z]).reshape(nat,3,3)
+        matrix = np.asarray([0.5*(matrix[i] + np.transpose(matrix[i])) for i in range(len(matrix))])
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: Since data is coming in as a list, I don't think `asarray` offers an advantage here over `array` (a copy will be required). You could try out something like `np.fromiter(0.5*(mat+np.transpose(mat)) for mat in matrix))`

File: exatomic/gaussian/output.py [hunk 8]
```diff
@@ -608,6 +619,212 @@ def parse_momatrix(self):
...
+            #print("====================={}=====================".format(i+1))
+            #print("Eigenvalues:\t{}\t{}\t{}".format(vals[0]*conv, vals[1]*conv, vals[2]*conv))
+            #print("Isotropic:\t{}\nAnisotropy:\t{}".format(iso[i]*conv, aniso[i]*conv))
+        matrix = matrix.reshape(len(matrix)*3, -1)
+        df = pd.DataFrame(matrix, columns=['x', 'y', 'z'])
+        label = 'nmr shielding'
+        # get atom center indexes
+        atom = [i for i in range(nat)]
+        # get atom symbols
+        symbols = list(map(lambda x: z2sym[x], znums))
```
Comment (avmarchenko) [inline exatomic/gaussian/output.py]: Since this will be passed to `np.tile`, there is no need to greedy evaluate the generator here. I would suggest removing the `list` as `np.tile` (below) will evaluate the generator (i.e. use the lazy evaluation mentality that generators are good for!).

File: exatomic/va/va.py [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+Collection of classes for VA program
+"""
+import numpy as np
+import pandas as pd
+import csv
+import os
+from exa.util.units import Length
+from exa.util.utility import mkp
+
+_gauss_template='''{link0}
```
Comment (avmarchenko) [inline exatomic/va/va.py]: This template is code-specific. I understand that the currently available input generator doesn't meet all of the needs but regardless of that I think we should keep code-specific code in the appropriate subpackage. What do you think about migrating the generation part to the `gaussian` subpackage?

File: exatomic/va/va.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+    position by using the displacement vector components contained in the
+    :class:`~exatomic.atom.Frequency` dataframe. It will scale these displacements to a
+    desired type defined by the user with the delta_type keyword. For more information
+    on this keyword see the documentation on the
+    :class:`~exatomic.va.va.GenInputs.gen_delta` function.
+
+    Args:
+        uni (:class:`~exatomic.Universe`): Universe object containg pertinent data
+        delta_type (int): Integer value to define the type of delta parameter to use
+    """
```
Comment (avmarchenko) [inline exatomic/va/va.py]: Very good documentation here. Thanks for being so descriptive!

File: exatomic/va/va.py [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+            # a = np.linalg.norm(freq[['dx', 'dy', 'dz']].
+            #                                    values, axis=1, ord=2)
+            # vec_sum = []
+            # for i in range(0,len(a),3):
+            #     vec_sum.append([0])
+            #     for j in range(i,i+3):
+            #         vec_sum[-1] += a[j]
+            # print(a)
+            # print(vec_sum)
+            d = freq.groupby(['freqdx', 'frame']).apply(
```
Comment (avmarchenko) [inline exatomic/va/va.py]: An alternative approach to this could be to do something like
```
@numba.jit
def func(args):
    return sum(linalg)

freq.groupby([...])[['dx', 'dy', 'dz']].apply(func)
```
which as highlight opens the possibility for performance gains on the computation itself (not sure if that is the bottleneck here).

File: exatomic/va/va.py [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+            freq_g = freq.groupby('freqdx').filter(lambda x: fdx in
+                                                    x['freqdx'].drop_duplicates().values).copy()
+        disp = freq_g[['dx','dy','dz']].values
+        modes = freq_g['frequency'].drop_duplicates().values
+        nat = len(eqcoord)
+        freqdx = freq_g['freqdx'].drop_duplicates().values
+        tnmodes = len(freq['freqdx'].drop_duplicates())
+        nmodes = len(freqdx)
+        # chop all values less than 1e-6
+        eqcoord[abs(eqcoord) < 1e-6] = 0.0
```
Comment (avmarchenko) [inline exatomic/va/va.py]: It could be worth making the tolerance a class attribute (e.g. `self._tol` or `self.tol`). Gives the "power" user more flexibility

File: exatomic/va/va.py [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+                    j.write(_sbatch.format(key="output", value=file.replace(".inp", ".out")))
+                    j.write(module)
+                    for line in f:
+                        j.write(line)
+                    j.write(end_com+'\n')
+                    f.close()
+                    j.close()
+
+    @staticmethod
+    def write_data_file(path, array, fn):
```
Comment (avmarchenko) [inline exatomic/va/va.py]: This seems generic enough it could be put elsewhere (or maybe in `exa` rather than here?).

File: exatomic/va/va.py [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+                    for key in sbatch.keys():
+                        j.write(_sbatch.format(key=key, value=sbatch[key])+'\n')
+                    j.write(_sbatch.format(key="job-name", value=file.replace(".inp", "")))
+                    j.write(_sbatch.format(key="output", value=file.replace(".inp", ".out")))
+                    j.write(module)
+                    for line in f:
+                        j.write(line)
+                    j.write(end_com+'\n')
+                    f.close()
+                    j.close()
```
Comment (avmarchenko) [inline exatomic/va/va.py]: When using context managers, you don't need to explicitly call `close` - it is done for you.

File: exatomic/va/va.py [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+                j_path = path+_name.format(job=job, int=fdx)
+            except OSError:
+                raise OSError("Failed to create directory {}".format(path+_name.format(
+                                                                                job=job, int=fdx)))
+            slurm = file.replace(".inp", ".slurm")
+            with open(path+file, 'r') as f:
+                with open(mkp(j_path, slurm), 'w') as j:
+                    j.write("#!/bin/bash\n")
+                    for key in sbatch.keys():
+                        j.write(_sbatch.format(key=key, value=sbatch[key])+'\n')
```
Comment (avmarchenko) [inline exatomic/va/va.py]: It might be better to generate the text and then do a single call to `j.write`.

File: exatomic/va/va.py [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+                p.write(_gauss_template.format(link0=link0, route=routep,
+                        title=str(fdx)+' property', charge=charge, mult=mult))
+                xyz['x'] *= Length['au', 'Angstrom']
+                xyz['y'] *= Length['au', 'Angstrom']
+                xyz['z'] *= Length['au', 'Angstrom']
+                xyz.to_csv(p, header=False, index=False, sep=' ', float_format=float_format,
+                            quoting=csv.QUOTE_NONE, escapechar=' ')
+                p.write('\n')
+
+    def gen_slurm_inputs(self, path, sbatch, module, end_com=''):
```
Comment (avmarchenko) [inline exatomic/va/va.py]: It seems that this method is generic enough to be place in a separate place. For example `exatomic.slurm.slurm`.

File: exatomic/core/gradient.py (hunk 2 above)
Comment (herbertludowieg) [reply exatomic/core/gradient.py]: Whoops I forgot to add that. I actually ended up adding it in the code for the parse_gradient. I'll make sure to add that in.

File: exatomic/gaussian/output.py (hunk 4 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: I'll look into it, but from what I understand with Gaussian the frequencies will only be written to the FChk file when we pass Freq=(SaveNormalModes) this will print out some data that have a very well defined number points. When it is not present it will not find anything for any of the keys. Again I'll make absolutely sure of this by running some small examples and see if I get any errors.

File: exatomic/gaussian/output.py (hunk 5 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: Would that cause any issues with the python 2.x release? I found something saying we may need to import from future?

File: exatomic/gaussian/output.py (hunk 6 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: I guess technically it could, but like I said before it should not execute that far as it will not find any of the needed frequency flags. I was trying to save a bit on IO by doing this, but it may be more stable to just find it separately. The only reason that I split up the functions was that I could not get them to work so that two class variables can be set within one function. Is this something that we could change? Because I know with the parse_gradients we could merge into the parse_atom function and save on some coding.

File: exatomic/gaussian/output.py (hunk 7 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: I'll try that. I know that I was doing this quite a bit and I must have forgotten to go back and change this one.

File: exatomic/gaussian/output.py (hunk 8 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: Yeah, that makes more sense I think I was just trying to use some of the things that were already being done that I know worked and went with it.

File: exatomic/va/va.py (hunk 9 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: So, like I said I did this a bit in a rush just to get something working as a draft and fine tune everything after. I was going back a lot on whether to put it in the specific quantum chemistry code package or have it here. Something I wanted to explore was to have a function where you specify the package you are using like adf or gaussian and then given the input it sends it to the respective input generator. At this moment all I can think of to do this is to have a lot of if statements which I think may be somewhat inefficient and finicky. I wanted to ask if you knew any better way to do this.

File: exatomic/va/va.py (hunk 10 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: Thanks I really wanted to make sure that people can use it and understand what the code is doing. I'm having that problem in transcribing the perl scripts :)

File: exatomic/va/va.py (hunk 11 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: The bottleneck is really the quantum chemistry calculations. This is being applied to a system with 121 atoms and there's not a big issue with this part of the code. It would benefit greatly if we could do less I/O iterations. It makes about 4*nmodes files which I have already worked on to only select a small list of modes.

File: exatomic/va/va.py (hunk 12 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: Yeah, I forgot to add it in here. Probably since I don't use it as much as in the gaussian.output.

File: exatomic/va/va.py (hunk 13 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: Yeah, but it's not going to be there for much longer. I have it in there just so I can compare to the existing program and see if what I am doing is good enough. It uses some FORTRAN scripts that I will be moving to python in the near future and then there'd be no need for the write_data_file, to_va, or write_grad_prop functions.

File: exatomic/va/va.py (hunk 14 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: I felt like that was the case, but I was just trying to be safe with it.

File: exatomic/va/va.py (hunk 15 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: Yeah, I should really do that.

File: exatomic/va/va.py (hunk 16 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: We could probably do that, but I feel like it won't be used that much since there are the job arrays in slurm that would be more efficient I think.

File: exatomic/va/va.py [hunk 17]
```diff
@@ -0,0 +1,327 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (herbertludowieg) [reply exatomic/va/va.py]: Definitely, I know that this is just the first portion and is in no way general enough to be fully implemented into the master branch and will be making numerous updates to it to get to the most general form that I can manage. Thank you for your inputs they're really helpful.

File: exatomic/core/gradient.py (hunk 2 above)
Comment (herbertludowieg) [reply exatomic/core/gradient.py]: I did add it in actually with the label column. I'm not sure if we should consider re-doing the columns to have adx be the atom index as opposed to label.

File: exatomic/gaussian/output.py (hunk 7 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: I ended up trying this
```
matrix = np.fromiter(0.5 * (mat + np.transpose(mat)) for mat in matrix)
```
But it needed a dtype parameter so I added that and [ ] for the for loop and it gave back the error
```
ValueError: setting an array element with a sequence.
```

File: exatomic/core/gradient.py (hunk 2 above)
Comment (avmarchenko) [reply exatomic/core/gradient.py]: Label isn't an index because an index must be a unique identifier. Labels repeat for universes that have more than one frame. If you look at the machinery of how universe dataframes are related (I think .network() will show that) you'll see which fields are indexes and their names.

File: exatomic/gaussian/output.py (hunk 5 above)
Comment (avmarchenko) [reply exatomic/gaussian/output.py]: Yeah you're right I think. My mistake.

File: exatomic/va/va.py (hunk 16 above)
Comment (avmarchenko) [reply exatomic/va/va.py]: Okay, I'm open to keeping it here. I don't want the scope of this pr to creep too much.

File: exatomic/va/va.py (hunk 13 above)
Comment (avmarchenko) [reply exatomic/va/va.py]: Okay understood.

File: exatomic/va/va.py (hunk 11 above)
Comment (avmarchenko) [reply exatomic/va/va.py]: Okay understood. Keep in mind that we want to try to not repeat ourselves (keep it DRY as they say). Even if performance isn't the main impetus, code cleanliness, testability, and atomicity could benefit.

File: exatomic/va/va.py (hunk 9 above)
Comment (avmarchenko) [reply exatomic/va/va.py]: I like the suggestion. I don't have the solution but its something that we should discuss. Maybe you can make your comment an issue and we can open up a broader discussion? For this PR, I guess lets not let the scope creep too much. One way to do that would be to put the template in the gaussian subpackage and make the template an argument in the cose below.

File: exatomic/gaussian/output.py (hunk 6 above)
Comment (avmarchenko) [reply exatomic/gaussian/output.py]: I think this comment deserves to out raised as an issue of the exa project. A minimal example would be good. I think you may have just highlighted a fundamental bug/limitation of the current 'typing' model of exa (Containers). Do you mind starting that issue up and I'll investigate more closely?

File: exatomic/gaussian/output.py (hunk 3 above)
Comment (tjduigna) [reply exatomic/gaussian/output.py]: Sounds like it boils down to downstream code either handling an `AttributeError` on the uni if `frequency_ext` doesn't exist or managing that `frequency_ext` is empty.  Not sure if there is an established precedent but a consistent one is a great idea.

File: exatomic/va/va.py (hunk 16 above)
Comment (tjduigna) [reply exatomic/va/va.py]: NB: Huge fan of a `exatomic.scheduler.slurm` concept.  At the very least templates and simple formatting.  The same applies for job array slurm scripts really.

I am curious if you looked at `exatomic.gaussian.inputs.functional_inputs` as an example of how to leverage the `Input.from_universe`  functionality, which is quite flexible for Gaussian.  If it was a restriction in the current functionality, it would be good to know where you ran into trouble to think about how to fix the design.

File: exatomic/va/va.py (hunk 13 above)
Comment (tjduigna) [reply exatomic/va/va.py]: 👍

File: exatomic/widgets/traits.py [hunk 18]
```diff
@@ -118,11 +118,17 @@ def tensor_traits(uni):
     except: idxs = [list(grp.index) for i, grp in grps]
     return {'tensor_d': grps.apply(lambda x: x.T.to_dict()).to_dict(), 'tensor_i': idxs}
 
+def freq_traits(uni):
+    grps = uni.frequency.groupby('frequency')
+    try: idxs = list(grps.groups.keys())
+    except: idxs = [list(grp.index) for i, grp in grps]
+    #print(idxs)
+    return {'freq_d': grps.apply(lambda x: x.T.to_dict()).to_dict(), 'freq_i': idxs}
```
Comment (tjduigna) [inline exatomic/widgets/traits.py]: Curious about the intentions for frequency traits.  Perhaps it might make a good issue to discuss the current state of frequency data and viz.

File: exatomic/gaussian/output.py (hunk 6 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: I guess my question is more why we can't set two class attributes within one function. Looking through the code every function sets one and I was trying to set self.frequency and self.frequency_ext within the parse_frequency function. Don't know if that makes more sense.

File: exatomic/va/va.py (hunk 14 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: I have taken out the file.close commands with this last push.

File: exatomic/widgets/traits.py (hunk 18 above)
Comment (herbertludowieg) [reply exatomic/widgets/traits.py]: It was something I was working on before that I thought would be interesting to implement I have just commented out all of the code I made for it since it doesn't work right now. Now with working on this vibrational averaging code I think it may be good to look more into it in the near future.

File: exatomic/va/va.py (hunk 11 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: I agree, I'll keep that in mind in the future.

File: exatomic/gaussian/output.py (hunk 6 above)
Comment (tjduigna) [reply exatomic/gaussian/output.py]: It does look like an issue.  There is an example of a workaround in the same file:
https://github.com/exa-analytics/exatomic/blob/6e4f62d43096a2fcf6d490489a2f8990808b133b/exatomic/gaussian/output.py#L267 
where `self.basis_set_order` is set inside of `parse_momatrix` and a dummy method `parse_basis_set_order` checks if it has been set and falls back to calling `parse_momatrix`.  Setting both attributes inside `parse_frequency` should in principle work.  I wonder if you are using the `to_universe` and the attribute is getting passed over when transferring attributes from the output to the uni.  This could just be a bug in that function, unless  of course, you aren't calling to_universe.

File: exatomic/gaussian/output.py [hunk 19]
```diff
@@ -409,12 +418,13 @@ def parse_frequency(self):
...
+        # TODO: Make absolutely sure what units gaussian reports the displacements as
         # TODO: verify with an external program that vibrational
         #       modes look the same as the ones generated with
         #       this methodology.
-        frequency['dx'] *= Length['Angstrom', 'au']
-        frequency['dy'] *= Length['Angstrom', 'au']
-        frequency['dz'] *= Length['Angstrom', 'au']
+        #frequency['dx'] *= Length['Angstrom', 'au']
+        #frequency['dy'] *= Length['Angstrom', 'au']
+        #frequency['dz'] *= Length['Angstrom', 'au']
```
[the remark this answers is not in the document]
Comment (tjduigna) [reply exatomic/gaussian/output.py]: Interesting, I never had the opportunity to validate this for myself so it is entirely possible that the displacements are in bohr already.

File: exatomic/widgets/traits.py (hunk 18 above)
Comment (tjduigna) [reply exatomic/widgets/traits.py]: Yeah it's likely the existing frequency code is among the most un-tested code in the library.  I definitely recommend continuing to vet what's there.

File: exatomic/gaussian/output.py (hunk 6 above)
Comment (avmarchenko) [reply exatomic/gaussian/output.py]: I think I still don't understand exactly the issue. Isn't the quoted code there an example of a method that sets two class attributes? Here is a minimal working example of setting two attributes in a single method. Note the change in type (due to the `TypedMeta` metaclass machinery.
```
from exa import Editor, TypedMeta
import six

class Meta(TypedMeta):
    atom = int
    frame = float
    
class Fchk(six.with_metaclass(Meta, Editor)):
    def parse_atom(self):
        self.atom = 1.0
        self.frame = 1
        
f = Fchk("", ignore=True)
f.atom, type(f.atom), f.frame, type(f.frame)    # (1, int, 1.0, float)
```

Does that help explain the behavior you are seeing?

File: exatomic/va/va.py (hunk 16 above)
Comment (avmarchenko) [reply exatomic/va/va.py]: What do you all think about putting that as an issue with the label feature request?

Comment (herbertludowieg) [reply exatomic/va/va.py]: I think we could do it. I guess I'm just worried about not making it too simple where a user ends up typing the entire file as an input. I'll try to think of some ideas to do this.

File: exatomic/core/gradient.py (hunk 2 above)
Comment (herbertludowieg) [reply exatomic/core/gradient.py]: replaced atom with label for the next push

File: exatomic/va/va.py (hunk 9 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: I opened up an issue and put some code I was working on that would be where I'm thinking about doing this.

File: exatomic/va/va.py (hunk 12 above)
Comment (herbertludowieg) [reply exatomic/va/va.py]: I added it in for the next push.

File: exatomic/gaussian/output.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: Yeah, I'll keep looking into it and change it back if it's in Angstroms instead

File: exatomic/widgets/traits.py (hunk 18 above)
Comment (herbertludowieg) [reply exatomic/widgets/traits.py]: Yeah, I'm just trying some things out. Will probably open up an issue in the coming months so I have time to make some concise ideas on what we could do with it.

File: exatomic/gaussian/output.py (hunk 3 above)
Comment (herbertludowieg) [reply exatomic/gaussian/output.py]: So then we'd have to create an empty container for every attribute? Only issue I can see is that if someone is trying to implement some code and is not aware of this they may have a harder time debugging as the error may be linked to their code as opposed to it saying this thing you are using does not exist.

**Comment by avmarchenko** [comment]
Is this ready for another review?

File: exatomic/va/va.py [hunk 20]
```diff
@@ -502,7 +502,89 @@ def __init__(self, uni, delta_type=0, fdx=-1, *args, **kwargs):
         self.delta = gen_delta(freq, delta_type)
         self.disp = self._gen_displaced(freq, atom, fdx)
 
+class VAMeta(TypedMeta):
+    grad_0 = Gradient
+    grad_plus = Gradient
+    grad_minus = Gradient
+    gradient = Gradient
+
+class VA(metaclass=VAMeta):
```
Comment (avmarchenko) [inline exatomic/va/va.py]: For py2 compatibility, you could use `six.with_metaclass(...)` (or something like that. That said, py2 is leaving Earth in a couple of years..

**Comment by avmarchenko** [comment]
I think the use of `TypedMeta` is all correct and as expected. I think it best to incorporate this PR and then iterate on its features/performance. I've marked it approved.
